### DISCUSSION TITLE
Compare an object with multiple objects at the same time.

### DISCUSSION THREAD

**Discussion by just-a-hriday** [dotnet/csharplang#5022] [Language Ideas]
Right now, there's no easy way to do an equality check on an object against multiple object using some sort of a loop, collection, or tons and tons of `x == a || x == b || x == c`.

Here's what I mean: 
```c#
// remember that this is just an example - there are better applications for my proposal.

// check if the name Henry is in the array of names.
string henry = "Henry";

// approach one:
Console.WriteLine(henry == "John" || henry == "Oliver" || henry ==  "George"...);

// approach two:
string[] names = new string[] { "John" , "Oliver" , "George" , "Harry" , "Henry" , "Theodore" , "William" , "Max" , "Tom" , "Benjamin" , "Jack" };
Console.WriteLine(names.Contains(henry));

// approach three:
string[] names = new string[] { "John" , "Oliver" , "George" , "Harry" , "Henry" , "Theodore" , "William" , "Max" , "Tom" , "Benjamin" , "Jack" };
foreach ( string name in names )
{
    Console.WriteLine(henry == name);
}
```

This is really annoying. To fix this, I'm proposing two new features:
```c#
// 1: an overload for object.Equals
// of course, this is assuming that LINQ is not allowed
bool Equals ( object objA , object[] objs )
{
    foreach ( object objB in objs )
    {
        if ( object.Equals(objA , objB) )
        {
            return true;
        }
    }

    return false;
}
```
This returns true if `objA` is equal to any element of `objs` - simply because if you were to return true only if `objA` matches all elements of `objs`, well, it would return false in all cases except for when all elements of `objs` are the same - in which case, the former would return the same.

```c#
// 2: new syntax
string henry = "Henry";
if ( henry == ( "John" || "Oliver" || "George" ) )
{
    Console.WriteLine($"{henry} is a valid name!");
}
```
---
For people pointing out that you can do
```c#
string henry = "Henry";
if ( henry is "John" or "Oliver" or "George" )
{
    Console.WriteLine($"{henry} is a valid name!");
}
```

That will only work for constant values - my suggestion is to make it possible for variables, too.
```c#
// this won't work
string name = "...";
static void Main ()
{
    if ( henry is Console.ReadLine() or name )
    {
        Console.WriteLine($"{henry} is a valid name!");
    }
}
```

**Comment by ufcpp** [comment]
```cs
var henry = "Henry";

if (henry is "John" or "Oliver" or "George")
{
    Console.WriteLine($"{henry} is a valid name!")
}
```

**Reply by just-a-hriday** [reply]
<strike>Woah, I didn't know that.</strike> No, that can only be used for constant values.

**Reply by just-a-hriday** [reply]
That can only compare `henry` to constant values - not variables. I'll edit my question to show that.

**Comment by koszeggy** [comment]
For constant expressions and if you can use C# 9.0 features you can use the enhanced pattern matching as @ufcpp mentioned:

```cs
if (name is "John" or "Oliver" or "George")
```

And for any other cases you can just write an extension method. I always liked the `in` operator in SQL and Pascal so I [created](https://github.com/koszeggy/KGySoft.CoreLibraries/blob/be14d6073688f74b4101ed9e34dc4233ef6282e7/KGySoft.CoreLibraries/CoreLibraries/_Extensions/ObjectExtensions.cs#L108) a simple C# version of it that can be used like this:

```cs
if (name.In("John", "Oliver", "George"))
```

**Reply by just-a-hriday** [reply]
The whole point of me suggesting this was to remove the need for extension methods.

**Reply by bernd5** [reply]
There is just one thing to mention - with such an extension method you loose the short-circuiting evaluation of operator ||.

**Reply by koszeggy** [reply]
@bernd5:
> with such an extension method you loose the short-circuiting evaluation of operator ||.

That's why it has an [overload](https://github.com/koszeggy/KGySoft.CoreLibraries/blob/be14d6073688f74b4101ed9e34dc4233ef6282e7/KGySoft.CoreLibraries/CoreLibraries/_Extensions/ObjectExtensions.cs#L155) where the elements of the `set` can be evaluated on demand. But as it is mentioned in the remarks, for constant expressions it is better to use the originally linked overload because delegates also have their own overhead.
